File: l2-domain/src/main/java/com/vvygulyarniy/l2/domain/geo/Point.java
package com.vvygulyarniy.l2.domain.geo;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.Objects;

import static java.lang.Math.round;
import static java.lang.Math.toIntExact;

/**
 * Phoen-X on 11.03.2017.
 */
@AllArgsConstructor
@Getter
public final class Point {
    private final int x;
    private final int y;
    private final int z;

    public double distanceTo(final Point other) {
        double x1 = x;
        double y1 = y;
        double z1 = z;

        double x2 = other.getX();
        double y2 = other.getY();
        double z2 = other.getZ();

        x1 -= x2;
        y1 -= y2;
        z1 -= z2;

        return Math.sqrt(x1 * x1 + y1 * y1 + z1 * z1);
    }

    public Point moveForwardTo(Point target, double distance) {
        double yLength = target.getY() - this.getY();
        double xLength = target.getX() - this.getX();
        double zLength = target.getZ() - this.getZ();

        double angleSin = yLength / this.distanceTo(target);
        double angleCos = xLength / this.distanceTo(target);
        double zAngleSin = zLength / this.distanceTo(target);

        double deltaX = distance * angleCos;
        double deltaY = distance * angleSin;
        double deltaZ = distance * zAngleSin;
        int newX = toIntExact(round(this.getX() + deltaX));
        int newY = toIntExact(round(this.getY() + deltaY));
        int newZ = toIntExact(round(this.getZ() + deltaZ));

        return new Point(newX, newY, newZ);
    }

    @Override
    public String toString() {
        return "(" + x + "," + y + "," + z + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Point point = (Point) o;
        return x == point.x && y == point.y && z == point.z;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y, z);
    }
}

File: l2-domain/src/main/java/com/vvygulyarniy/l2/domain/geo/Point.kt
package com.vvygulyarniy.l2.domain.geo

import java.lang.Math.round
import java.lang.Math.toIntExact

data class Point(val x: Int = 0, val y: Int = 0, val z: Int = 0) {

    fun distanceTo(other: Point): Double {
        var x1 = x.toDouble()
        var y1 = y.toDouble()
        var z1 = z.toDouble()

        val x2 = other.x.toDouble()
        val y2 = other.y.toDouble()
        val z2 = other.z.toDouble()

        x1 -= x2
        y1 -= y2
        z1 -= z2

        return Math.sqrt(x1 * x1 + y1 * y1 + z1 * z1)
    }

    fun moveForwardTo(target: Point, distance: Double): Point {
        val yLength = (target.y - this.y).toDouble()
        val xLength = (target.x - this.x).toDouble()
        val zLength = (target.z - this.z).toDouble()

        val angleSin = yLength / this.distanceTo(target)
        val angleCos = xLength / this.distanceTo(target)
        val zAngleSin = zLength / this.distanceTo(target)

        val deltaX = distance * angleCos
        val deltaY = distance * angleSin
        val deltaZ = distance * zAngleSin
        val newX = toIntExact(round(this.x + deltaX))
        val newY = toIntExact(round(this.y + deltaY))
        val newZ = toIntExact(round(this.z + deltaZ))

        return Point(newX, newY, newZ)
    }

    override fun toString(): String {
        return "($x,$y,$z)"
    }
}
